scipy/spatial/transform/rotation.py

```diff
@@ -0,0 +1,862 @@
+            dcm_transformed[~safe2, 0, 0] - dcm_transformed[~safe2, 1, 1]
+        )
+    else:
+        # For instrinsic, set third angle to zero
+        # 6a
+        angles[~safe_mask, 2] = 0
+        # 6b
+        angles[~safe1, 0] = np.arctan2(
+            dcm_transformed[~safe1, 1, 0] - dcm_transformed[~safe1, 0, 1],
+            dcm_transformed[~safe1, 0, 0] + dcm_transformed[~safe1, 1, 1]
+        )
+        # 6c
+        angles[~safe2, 0] = np.arctan2(
+            dcm_transformed[~safe2, 1, 0] + dcm_transformed[~safe2, 0, 1],
+            dcm_transformed[~safe2, 0, 0] - dcm_transformed[~safe2, 1, 1]
+        )
+
+    # Step 7
+    if seq[0] == seq[2]:
+        # lambda = 0, so we can only ensure angle2 -> [0, pi]
+        adjust_mask = np.logical_or(angles[:, 1] < 0, angles[:, 1] > np.pi)
+    else:
+        # lambda = + or - pi/2, so we can ensure angle2 -> [-pi/2, pi/2]
+        adjust_mask = np.logical_or(angles[:, 1] < -np.pi / 2,
+                                    angles[:, 1] > np.pi / 2)
+
+    # Dont adjust gimbal locked angle sequences
+    adjust_mask = np.logical_and(adjust_mask, safe_mask)
+
+    angles[adjust_mask, 0] += np.pi
+    angles[adjust_mask, 1] = 2 * offset - angles[adjust_mask, 1]
+    angles[adjust_mask, 2] -= np.pi
+
+    angles[angles < -np.pi] += 2 * np.pi
+    angles[angles > np.pi] -= 2 * np.pi
+
+    # Step 8
+    if not np.all(safe_mask):
+        warnings.warn("Gimbal lock detected. Setting third angle to zero since"
+                      " it is not possible to uniquely determine all angles.")
+
+    # Reverse role of extrinsic and intrinsic rotations, but let third angle be
+    # zero for gimbal locked cases
+    if extrinsic:
+        angles = angles[:, ::-1]
+    return angles
+
+
+def _make_elementary_quat(axis, angles):
+    num_rotations = angles.shape[0]
+    quat = np.zeros((num_rotations, 4))
+
+    quat[:, 3] = np.cos(angles / 2)
+    quat[:, AXIS_TO_IND[axis]] = np.sin(angles / 2)
+    return quat
+
+
+def _compose_quat(p, q):
+    product = np.empty((max(p.shape[0], q.shape[0]), 4))
+    # Scalar part of result
+    product[:, 3] = p[:, 3] * q[:, 3] - np.sum(p[:, :3] * q[:, :3], axis=1)
+    # Vector part of result
+    product[:, :3] = (p[:, None, 3] * q[:, :3] + q[:, None, 3] * p[:, :3] +
+                      np.cross(p[:, :3], q[:, :3]))
+    return product
+
+
+def _elementary_quat_compose(seq, angles, intrinsic=False):
+    # Initialize result to first axis
+    result = _make_elementary_quat(seq[0], angles[:, 0])
+
+    for idx, axis in enumerate(seq[1:], start=1):
+        if intrinsic:
+            result = _compose_quat(
+                result,
+                _make_elementary_quat(axis, angles[:, idx]))
+        else:
+            result = _compose_quat(
+                _make_elementary_quat(axis, angles[:, idx]),
+                result)
+    return result
+
+
+class Rotation(object):
+    """Rotation transform in 3 dimensions.
```

Comment:
I guess it is usually said as "Rotation transformation" or you can just put "Rotation in 3 dimensions".